source/agora/node/NodeTest.d

```diff
@@ -0,0 +1,364 @@
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.NodeTest;
+
+import agora.node.Node;
+import agora.common.API;
+import agora.common.Config;
+import agora.common.crypto.Key;
+import agora.common.Version;
+import agora.node.Network;
+
+import vibe.d;
+
+import std.array;
+import std.algorithm;
+import std.exception;
+
+/*******************************************************************************
+
+    Node registry / set
+
+    This interface exposes basic functions to add, remove, and contact nodes.
+
+    Each node maintains and requires a registry of nodes they know about.
+    Among those nodes, some might not be trusted, some might be disabled, etc...
+    As the result, the set of nodes in this registry might not be a quorum,
+    but a larger set. It might also be a quorum, or a slice of it.
+    Modifications of a set might affect dependent sets.
+
+*******************************************************************************/
+
+public interface INodeSet
+{
+@safe:
+
+    /// Add the key & address to the node set
+    public void add (PublicKey key, string address);
+
+    /// Remove the key from the set
+    public void remove (PublicKey key);
+
+    /// Get the API of the specified key (or null if it doesn't exist)
+    public API getAPI (PublicKey key);
+
+    /// Get the array of all APIs
+    public API[] getAllAPIs ( );
+}
+
+///
+public abstract class TestRegistry : INodeSet
+{
+@safe:
+
+    /// Map of pubkey to nodes
+    protected API[PublicKey] registry;
+
+
+    /***************************************************************************
+
+        Factory method, map node names to instances
+
+        Should either `assert` or `throw` if the identifier can't be matched,
+        it should never return `null`.
+
+        Params:
+            key  = The key to use for this node
+            name = Identifier describind the node to instantiate
+
+        Returns:
+            An non-`null` instance of an `API`
+
+    ***************************************************************************/
+
+    protected abstract API factory (Config config, string name);
+
+    ///
+    public void register (Config config, string name)
+    {
+        this.registry[config.node.key_pair.address] = this.factory(config, name);
+    }
+
+    ///
+    public override void add (PublicKey key, string address)
+    {
+        assert(key in this.registry);
+    }
+
+    ///
+    public override void remove (PublicKey key)
+    {
+    }
+
+    ///
+    public override API getAPI (PublicKey key)
+    {
+        if (auto api = key in this.registry)
+            return *api;
+
+        return null;
+    }
+
+    ///
+    public override API[] getAllAPIs ( )
+    {
+        return this.registry.byValue.array;
+    }
+}
+
+///
+class TestNodeRegistry : TestRegistry
```

Comment:
For the moment let's keep it as it, but the separation was done so that you could have different registry for different tests (different registry == different node behavior)

Reply:
Btw, how exactly do we plan on simulating a "bad" node? In theory you can only interact with it through its publicly exposed API.

Reply:
I'm guessing through subclassing the Node class somehow?

Reply:
Yes

Reply:
You subclass the node class and implement the behavior. Then in `registry`, you give it a name, e.g. `byzantine` or `dead` and return the appropriate node.

Reply:
Ah, I see now why you have that name string passed as a parameter. Got it.